src/t_zset.c

```diff
@@ -2762,73 +2745,73 @@ void zunionInterDiffGenericCommand(client *c, robj *dstkey, int numkeysIndex, in
                 /* If we don't have it, we need to create a new entry. */
-                if (!existing) {
-                    tmp = zuiNewSdsFromValue(&zval);
+                void *existing;
+                if (hashtableFindPositionForInsert(temp_ht, sdsval, &position, &existing)) {
+                    struct sdsDoublePair *new_element = zmalloc(sizeof(struct sdsDoublePair));
+                    new_element->key = zuiNewSdsFromValue(&zval);
+                    new_element->score = score;
+                    hashtableInsertAtPosition(temp_ht, new_element, &position);
                     /* Remember the longest single element encountered,
                      * to understand if it's possible to convert to listpack
                      * at the end. */
-                    totelelen += sdslen(tmp);
-                    if (sdslen(tmp) > maxelelen) maxelelen = sdslen(tmp);
-                    /* Update the element with its initial score. */
-                    dictSetKey(dstzset->dict, de, tmp);
-                    dictSetDoubleVal(de, score);
+                    totelelen += sdslen(new_element->key);
+                    if (sdslen(new_element->key) > maxelelen) {
+                        maxelelen = sdslen(new_element->key);
+                    }
                 } else {
                     /* Update the score with the score of the new instance
-                     * of the element found in the current sorted set.
-                     *
-                     * Here we access directly the dictEntry double
-                     * value inside the union as it is a big speedup
-                     * compared to using the getDouble/setDouble API. */
-                    double *existing_score_ptr = dictGetDoubleValPtr(existing);
-                    zunionInterAggregate(existing_score_ptr, score, aggregate);
+                     * of the element found in the current sorted set. */
+                    struct sdsDoublePair *pair = existing;
+                    zunionInterAggregate(&pair->score, score, aggregate);
                 }
             }
             zuiClearIterator(&src[i]);
         }
 
         /* Step 2: convert the dictionary into the final sorted set. */
-        di = dictGetIterator(dstzset->dict);
-
-        while ((de = dictNext(di)) != NULL) {
-            sds ele = dictGetKey(de);
-            score = dictGetDoubleVal(de);
-            znode = zslInsert(dstzset->zsl, score, ele);
-            dictSetVal(dstzset->dict, de, &znode->score);
+        hashtableExpand(dstzset->ht, hashtableSize(temp_ht));
+        hashtableIterator iter;
+        hashtableInitIterator(&iter, temp_ht);
+
+        void *next;
+        while (hashtableNext(&iter, &next)) {
+            struct sdsDoublePair *element = next;
+            zskiplistNode *znode = zslInsert(dstzset->zsl, element->score, element->key);
+            element->key = NULL; /* passed ownership to skiplist node */
+            hashtableAdd(dstzset->ht, znode);
```

Comment:
I feel uncomfortable with this part. Did we check the perf impact of this copy operation, unions can sometimes introduce sets with millions of elements. and we are also using twice the hashtable memory for maintaining this. 
I am not sure why we can't just create the skiplist nodes (based on random heights matching the current expansion) and insert them to the skiplist after we complete and scan the hashtable.
Or maybe I missed something?

Reply:
I missed that updating the scores is much cheaper if we haven't yet built the skiplist. Ran, your idea sounds good, if I understand it correctly. In step 1, should we create skiplist nodes without initializing the pointers? Then in step 2 we link the nodes together? We'd have to refactor zslInsert to separate the node creation and the insertion into the skiplist?

Reply:
yes that is basically what I meant

Reply:
I guess I'm not sure what you mean when you say based on random heights, but I think I get the idea. Yes, that would be more efficient. I'll take a look at making this change.

Reply:
> I guess I'm not sure what you mean when you say based on random heights

When a skiplist node is created, it always gets a random height, i.e. a random number of levels with pointers to other nodes. So, just "create a skiplist node" implies it has some random height.